File: Command/c++/Command.cpp
#include <iostream>
#include <vector>
#include <stack>

class ICommand
{
    public:
        virtual void positive() = 0;
        virtual void negative() = 0;
};

class Conveyor
{
    public:
        void on();
        void off();
        void speedIncrease();
        void speedDecrease();
};

class ConveyorWorkCommand : public ICommand
{
    private:
        Conveyor* conveyor;

    public:
        ConveyorWorkCommand(Conveyor* c) : conveyor(c) {}
        void positive() override;
        void negative() override;
};

class ConveyorAdjustCommand : public ICommand
{
    private:
        Conveyor* conveyor;

    public:
        ConveyorAdjustCommand(Conveyor* c) : conveyor(c) {}
        void positive() override;
        void negative() override;
};

class Multipult
{
    private:
        std::vector<ICommand*> commands;
        std::stack<ICommand*> history;
        
    public:
        Multipult();
        void SetCommand(int button, ICommand* command);
        void pressOn(int button);
        void pressCancel();
};

int main()
{
    Conveyor* conveyor = new Conveyor();
    
    Multipult* multipult = new Multipult();
    multipult->SetCommand(0, new ConveyorWorkCommand(conveyor));
    multipult->SetCommand(1, new ConveyorAdjustCommand(conveyor));

    multipult->pressOn(0);
    multipult->pressOn(1);
    multipult->pressCancel();
    multipult->pressCancel();
    return 0;
}

void Conveyor::on()
{
    std::cout << "the pipeline is running\n";
}

void Conveyor::off()
{
    std::cout << "the conveyor is stopped\n";
}

void Conveyor::speedIncrease()
{
    std::cout << "increased conveyor speed\n";
}

void Conveyor::speedDecrease()
{
    std::cout << "reduced conveyor speed\n";
}

void ConveyorWorkCommand::positive()
{
    conveyor->on();
}

void ConveyorWorkCommand::negative()
{
    conveyor->off();
}

void ConveyorAdjustCommand::positive()
{
    conveyor->speedIncrease();
}

void ConveyorAdjustCommand::negative()
{
    conveyor->speedDecrease();
}

Multipult::Multipult()
{
    commands.resize(2);
}

void Multipult::SetCommand(int button, ICommand *command)
{
    commands[button] = command;
}

void Multipult::pressOn(int button)
{
    commands[button]->positive();
    history.push(commands[button]);
}

void Multipult::pressCancel()
{
    if (!history.empty())
    {
        history.top()->negative();
        history.pop();
    }
}

File: Command/java/Command.java
import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

interface ICommand{
    void positive();
    void negative();
}

class Conveyor{
    public void on(){
        System.out.println("the pipeline is running");
    }

    public void off(){
        System.out.println("the conveyor is stopped");
    }

    public void speedIncrease(){
        System.out.println("increased conveyor speed");
    }
    
    public void speedDecrease(){
        System.out.println("reduced conveyor speed");
    }
}

class ConveyorWorkCommand implements ICommand{
    private Conveyor conveyor;

    public ConveyorWorkCommand(Conveyor conveyor){
        this.conveyor = conveyor;
    }

    @Override
    public void positive(){
        conveyor.on();
    }

    @Override
    public void negative(){
        conveyor.off();
    }
}

class ConveyorAdjustCommand implements ICommand{
    private Conveyor conveyor;

    public ConveyorAdjustCommand(Conveyor conveyor){
        this.conveyor = conveyor;
    }

    @Override
    public void positive(){
        conveyor.speedIncrease();
    }

    @Override
    public void negative(){
        conveyor.speedDecrease();
    }
}

class MultiPult{
    private List<ICommand> commands;
    private Stack<ICommand> history;

    public MultiPult(){
        commands = new ArrayList<>();
        commands.add(null);
        commands.add(null);
        history = new Stack<>();
    }

    public void setCommand(int button, ICommand command){
        commands.set(button, command);
    }

    public void pressON(int button){
        commands.get(button).positive();
        history.push(commands.get(button));
    }

    public void pressCancel(){
        if (!history.empty()){
            history.pop().negative();
        }
    }
}

public class Main {
    public static void main(String[] args){

        Conveyor conveyor = new Conveyor();

        MultiPult multiPult = new MultiPult();
        multiPult.setCommand(0, new ConveyorWorkCommand(conveyor));
        multiPult.setCommand(1, new ConveyorAdjustCommand(conveyor));

        multiPult.pressON(0);
        multiPult.pressON(1);
        multiPult.pressCancel();
        multiPult.pressCancel();
    }
}